File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/Array_List/DoublePointer/List/mergeKLists_23.java
package algo.tzashinorpu.SpecialSubject.Array_List.DoublePointer.List;

import algo.tzashinorpu.SpecialSubject.ListNode;

import java.util.Comparator;
import java.util.PriorityQueue;

public class mergeKLists_23 {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode dummyHead = new ListNode();
        ListNode p = dummyHead;
        PriorityQueue<ListNode> queue = new PriorityQueue<>(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                return Integer.compare(o1.val, o2.val);
            }
        });
        for (ListNode l : lists) {
            if (l != null) queue.add(l);
        }
        while (queue.size() > 0) {
            ListNode poll = queue.poll();
            p.next = poll;
            p = p.next;
            if (poll.next != null) queue.add(poll.next);
        }
        return dummyHead.next;
    }
}

File: Clion/labuladong/chapter0/linklist/mergeKLists_23.cpp
//
// Created by TzashiNorpu on 5/31/2023.
//
#include "../../header/header.h"

using namespace labuladong;

ListNode *Solution::mergeKLists(vector<ListNode *> &lists) {
  struct compare {
      bool operator()(const ListNode *l, const ListNode *r) {
        return l->val > r->val;
      }
  };
  priority_queue < ListNode * , vector < ListNode * >, compare > q;
  for (auto l: lists) {
    if (l)q.push(l);
  }
  auto dummy = new ListNode(-1);
  auto x = dummy;
  while (!q.empty()) {
    auto top = q.top();
    q.pop();
    x->next = top;
    if (top->next) q.push(top->next);
    x = x->next;
  }
  return dummy->next;
}